File: su/2026-05/05_pgr_디펜스게임.py
import heapq 

def solution(n, k, enemy):
    answer = 0
    total_len = len(enemy)
    enemy_hq = []
    heapq.heapify(enemy_hq)
    
    for i in range(total_len):
        if n < enemy[i]:  # 무적권 써야 하는 경우
            if k == 0:  # 더이상 무적권 없는 경우
                return i
            
            heapq.heappush(enemy_hq, (-enemy[i], i))
            max_value, max_idx = heapq.heappop(enemy_hq)
            max_value = abs(max_value)  # 양수로 변환
            n += max_value  # 무적권 쓴 거 플러스
            n -= enemy[i]  # 현재 적군 막기
            k -= 1
            

        else:  # 현재 있는 병사로 막을 수 있는 경우
            heapq.heappush(enemy_hq, (-enemy[i], i))
            n -= enemy[i]
            
    answer = i + 1
    return answer

print(solution(2, 4, [3, 3, 3, 3]))
# print(solution(7, 3, [4, 2, 4, 5, 3, 3, 1]))

File: su/2026-05/05_pgr_디펜스게임.java
import java.util.*;

class Solution {
    public int solution(int n, int k, int[] enemy) {
        int answer = 0;
        int totalLen = enemy.length;

        // 최대 힙
        PriorityQueue<Integer> enemyPq = new PriorityQueue<>(Collections.reverseOrder());

        for (int i = 0; i < totalLen; i++) {
            if (n < enemy[i]) { // 무적권 써야 하는 경우
                if (k == 0) { // 더 이상 무적권 없는 경우
                    return i;
                }

                enemyPq.offer(enemy[i]);

                int maxValue = enemyPq.poll(); // 지금까지 나온 적 중 가장 큰 값
                n += maxValue;                 // 그 라운드는 무적권 쓴 것으로 처리해서 병사 환급
                n -= enemy[i];                 // 현재 라운드 적 처리
                k--;

            } else { // 현재 병사로 막을 수 있는 경우
                enemyPq.offer(enemy[i]);
                n -= enemy[i];
            }
        }

        answer = totalLen;
        return answer;
    }
}